### PR TITLE
Simplify `ArrayReprImpl`

### PR DESCRIPTION
**Pull request by asi1024**
This PR is ready for #6452.

### CODE REVIEW

File: chainerx_cc/chainerx/array_repr.cc [hunk 1]
```diff
@@ -257,6 +233,7 @@ std::ostream& operator<<(std::ostream& os, const Array& array) {
     // TODO(hvy): We need to determine the output specification of this function, whether or not to align with Python repr specification,
     // and also whether this functionality should be defined in C++ layer or Python layer.
     // TODO(hvy): Consider using a static dimensionality.
+    NoBackpropModeScope scope{};
```
Comment (okapies) [inline chainerx_cc/chainerx/array_repr.cc]: I'd like to just make sure that it is not related to the scope of this PR or not?

Comment (asi1024) [reply chainerx_cc/chainerx/array_repr.cc]: This fix is related to this PR.
https://github.com/chainer/chainer/blob/926e3f9fc058831191eee009d9ffe2d324371b82/chainerx_cc/chainerx/array_repr.cc#L224
Current implementation calls `internal::At` which is one of the routines, but we don't want create backward graph.

Comment (okapies) [reply chainerx_cc/chainerx/array_repr.cc]: Makes sense. Thanks for the clarification.

**Comment by asi1024** [comment]
@okapies Could you take a look?

**Comment by okapies** [comment]
Jenkins, test this please.

**Comment by okapies** [comment]
Jenkins, test this please

File: chainerx_cc/chainerx/array_repr.cc [hunk 2]
```diff
@@ -238,16 +200,30 @@ struct ArrayReprImpl {
...
+        if (ndim == 0) {
+            formatter.Print(os, static_cast<T>(AsScalar(array)));
+            return;
+        }
+        os << "[";
+        int64_t size = array.shape().front();
+        for (int64_t i = 0; i < size; ++i) {
+            if (i != 0) {
+                os << ",";
+                if (ndim > 1 || i % kMaxItemNumPerLine == 0) {
```
Comment (okapies) [inline chainerx_cc/chainerx/array_repr.cc]: I noticed that `ndim` is always `> 1` because `ndim != 0` in this place.

File: chainerx_cc/chainerx/array_repr.cc [hunk 3]
```diff
@@ -238,16 +200,30 @@ struct ArrayReprImpl {
...
-        IndexableArray<const T, Ndim> iarray{array};
+    static constexpr int kMaxItemNumPerLine = 10;
 
-        for (auto it = indexer.It(0); it; ++it) {
-            visitor(iarray, it);
+    template <typename T>
+    void ArrayReprRecursive(const Array& array, Formatter<T>& formatter, size_t indent, std::ostream& os) const {
+        const uint8_t ndim = array.ndim();
+        if (ndim == 0) {
+            formatter.Print(os, static_cast<T>(AsScalar(array)));
```
Comment (okapies) [inline chainerx_cc/chainerx/array_repr.cc]: It seems you changed to use `static_cast` instead of `reinterpret_cast` which is the implementation of `native::StorageToDataType<const T>`. Is it intended fix? https://github.com/chainer/chainer/blob/ca48f1b8efbf980423631accb8264bd8bdfa8ccd/chainerx_cc/chainerx/native/data_type.h#L16-L18

Comment (asi1024) [reply chainerx_cc/chainerx/array_repr.cc]: The original implementation uses `iarray`: `IndexableArray`, but I use `AsScalar` because `array` is guaranteed to be 0-dim array here.  `AsScalar` casts the data from `StorageType` to `DataType` internally, so we don't have to use `reinterpret_cast` here.
https://github.com/chainer/chainer/blob/b619124e1d88ce50a015c2a508184e843398f738/chainerx_cc/chainerx/routines/manipulation.cc#L48

**Comment by okapies** [comment]
Jenkins, test this please.

**Comment by okapies** [comment]
Jenkins, test this please.

**Comment by okapies** [comment]
Jenkins, test this please.
